src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,348 @@
+        subTlvCur = subTlvCur->GetNext();
+        VerifyOrExit(queryId->IsValid(), error = OT_ERROR_PARSE);
+    }
+    else
+    {
+        ExitNow(error = OT_ERROR_PARSE);
+    }
+
+    // Link Metrics Report TLV
+    tlv.SetType(Mle::Tlv::kLinkMetricsReport);
+    SuccessOrExit(error = aMessage.Append(&tlv, sizeof(tlv)));
+
+    if (queryId->GetSeriesId() == 0)
+    {
+        // Parse QueryOptions Sub-TLV
+        if (subTlvCur->GetType() == kLinkMetricsQueryOptions)
+        {
+            queryOptions = static_cast<const LinkMetricsQueryOptions *>(subTlvCur);
+            VerifyOrExit(queryOptions->IsValid(), error = OT_ERROR_PARSE);
+            subTlvCur = subTlvCur->GetNext();
+            VerifyOrExit(subTlvCur == subTlvEnd, error = OT_ERROR_PARSE);
+        }
+        else
+        {
+            ExitNow(error = OT_ERROR_PARSE);
+        }
+        SuccessOrExit(error = AppendSingleProbeLinkMetricsReport(aMessage, length, queryOptions,
+                                                                 Get<Mac::Mac>().GetNoiseFloor(), aRequestMessage));
+    }
+    else
+    {
+        ExitNow(error = OT_ERROR_NOT_IMPLEMENTED);
+    }
+
+    tlv.SetLength(length);
+    aMessage.Write(startOffset, sizeof(tlv), &tlv);
+
+exit:
+    otLogDebgMle("AppendLinkMetricsReport, error:%d", error);
+    return error;
+}
+
+void LinkMetrics::HandleLinkMetricsReport(const Message &     aMessage,
+                                          uint16_t            aOffset,
+                                          uint16_t            aLength,
+                                          const Ip6::Address &aAddress)
+{
+    otLinkMetricsValues    metricsValues;
+    uint16_t               pos    = aOffset;
+    uint16_t               endPos = aOffset + aLength;
+    Tlv                    tlv;
+    LinkMetricsTypeIdFlags typeIdFlags;
+
+    VerifyOrExit(mLinkMetricsReportCallback != nullptr, OT_NOOP);
+
+    memset(&metricsValues, 0, sizeof(otLinkMetricsValues));
+
+    otLogDebgMle("Received Link Metrics Report");
+
+    while (pos < endPos)
+    {
+        aMessage.Read(pos, sizeof(Tlv), &tlv);
+        VerifyOrExit(tlv.GetType() == kLinkMetricsReportSub, OT_NOOP);
+        pos += sizeof(Tlv);
+        VerifyOrExit(pos + tlv.GetLength() <= endPos, OT_NOOP);
+
+        aMessage.Read(pos, sizeof(LinkMetricsTypeIdFlags), &typeIdFlags);
+        VerifyOrExit(!typeIdFlags.IsExtendedFlagSet(), OT_NOOP); ///< Extended Flag is not used in Thread 1.2
+        pos += sizeof(LinkMetricsTypeIdFlags);
+
+        switch (typeIdFlags.GetMetricEnum())
+        {
+        case OT_LINK_METRICS_PDU_COUNT:
+            VerifyOrExit(typeIdFlags.IsLengthFlagSet(), OT_NOOP);
+            metricsValues.mMetrics.mPduCount = true;
+            aMessage.Read(pos, sizeof(uint32_t), &metricsValues.mPduCountValue);
+            pos += sizeof(uint32_t);
+            otLogDebgMle(" - PDU Counter: %d %s", metricsValues.mPduCountValue,
```

Comment:
What about moving the metric names (PDU Counter, LQI, Link Margin, RSSI) to const strings as well? They are also used in `cli.cpp`.

Reply:
Added dB and dBm in the string. 

I think adding const strings for the names won't save a lot of bytes(as they are short) and currently the log is more intuitive. Besides, these only matter when Debug log is enabled. I think we can refine it when we really need to exploit these bytes.

Reply:
I'd say saving some bytes it's always good, and it's easier to do so when you are working in this part of the code than in the future when first you need to identify the opportunity and then apply the changes without breaking anything.

In any case, not a big blocker from my side to move forward this PR.